### PR TITLE
Add guide on reassembling multi-segment messages

### PR DESCRIPTION
**Pull request by zedoraps**
## 🔍 Problem

- Some log sources cap transport-message size and split a single logical event into several numbered segments sharing an id.
- Segments arrive interleaved and ids get reused over time, so the partial messages are useless until reassembled.
- TQL had no guide documenting how to put such segments back together.

## 🛠️ Solution

- New generic guide `guides/parsing/reassemble-multi-segment-messages` (sidebar: Guides → Parsing).
- Documents the pattern: frame segments with `parse_grok`, gather them with an outer `window` and an inner `group` keyed on the id, then `sort` + `collect` + `first` to concatenate payloads in order.
- `window` outermost (not `group`) so state is bounded to the open windows rather than one subpipeline per near-unique id; the note explains why and when to invert.
- Points at work-with-time for supplying an event-time field, and at the `cisco::ise::*` operators as a real-world implementation.

## 💬 Review

- `window`-outer vs `group`-outer equivalence verified by running both on interleaved sample data; output is identical.
- Example pipelines verified to run; `markdownlint` passes on the changed file.

<sub>
⚙️ Code PR: tenzir/library#161
</sub>

### CODE REVIEW

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+segment it is. A <Fn>parse_grok</Fn> pattern pulls the shared message id, the
+segment total, the segment number, and the payload that follows. Keep the values
+as strings with `raw=true` so the parser doesn't reinterpret identifiers:
+
+```tql
+let $header = r"^(?:(?<category>CISE_\S+)\s+)?(?<message_id>\d+)\s+(?<segment_total>\d+)\s+(?<segment_number>\d+)\s+(?<payload>.*)$"
+
+from {content: r"CISE_Failed_Attempts 0001133664 4 0 ... Failed-Attempt, UserName=alice"},
+     {content: r"CISE_Failed_Attempts 0001133664 4 1  SSID=AA-BB, Step=11001"}
+segment = content.parse_grok($header, raw=true)
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Fixed in 5fcfeec. The framing step now notes the counter stays a string, and the gather example casts it with `segment_number = int(segment_number)` before `sort`, so segment 10 sorts after 2.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+Inside the window, <Op>sort</Op> the segments by number and collect their
+payloads with <Fn>collect</Fn>, which preserves the sorted order. Carry the
+message-level fields through with <Fn>first</Fn>:
+
+```tql
+from {message_id: "1", segment_total: "3", segment_number: 2, ts: 2024-01-01T10:00:00, payload: "c=3"},
+     {message_id: "1", segment_total: "3", segment_number: 0, ts: 2024-01-01T10:00:00, payload: "a=1"},
+     {message_id: "1", segment_total: "3", segment_number: 1, ts: 2024-01-01T10:00:00, payload: "b=2"}
+group message_id {
+  window size=30s, on=ts {
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Fixed. The example was reworked to put `window` on the outside (an inner `group` over a near-unique id would leak one subpipeline per message) and now sets `idle_timeout` on every windowed example, so the last message for each id flushes on low-volume sources.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+## Put it together
+
+The complete pipeline reads ISE syslog over TCP, reconstructs event time,
+reassembles the segments, and then parses and maps the result. The Cisco package
+ships these steps as the reusable operators `cisco::ise::reassemble` and
+`cisco::ise::parse`:
+
+```tql
+from_tcp "0.0.0.0:514" {
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Resolved in the docs: the generic rewrite dropped the full `from_tcp`/syslog pipeline, so the guide no longer shows that operator. Good catch on the semantics though — the package example `cisco/examples/ise-syslog-to-ocsf.tql` still uses `from_tcp "0.0.0.0:514"` and will be switched to `accept_tcp` in the code PR (tenzir/library#161).

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+Adjust the framing pattern, the identifier field, and the window size to match
+your source's segmentation scheme.
+
+## See Also
+
+- <Op>group</Op>
+- <Op>window</Op>
+- <Op>sort</Op>
+- <Fn>collect</Fn>
+- <Fn>parse_grok</Fn>
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Added reciprocal See Also links to `window` and `group`, where reassembly is a notable application. Left `sort`, `collect`, and `parse_grok` out on purpose: they're general-purpose primitives and a niche reassembly guide isn't a representative use, so linking it there would be noise.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+gather the segments of one message in event-time windows, and concatenate them
+into one record—then harden the pipeline against out-of-order arrival, duplicate
+delivery, and missing segments.
+
+## The problem
+
+Many log sources cap the size of a single transport message. When an event
+exceeds that cap, the source splits it into several smaller messages that share
+an identifier and carry a segment counter. The receiver sees a burst of partial
+messages that mean nothing on their own—you have to put them back together
```
Comment (mavam) [inline src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Second em dash; maybe add prompt to reduce those.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+This guide shows you how to reassemble a single logical message that a source
+split across several transport messages. You'll learn to frame each segment,
+gather the segments of one message in event-time windows, and concatenate them
+into one record—then harden the pipeline against out-of-order arrival, duplicate
+delivery, and missing segments.
+
+## The problem
+
+Many log sources cap the size of a single transport message. When an event
```
Comment (mavam) [inline src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx:12]: Would be nice to have few vendor/product names in parentheses here after "Many log sources".

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+before you can parse the event.
+
+A segmented message typically frames each part with a shared id, the total number
+of segments, and this segment's number, followed by the payload:
+
+```text
+f47ac10b-58cc 3 0 first part of the body
+f47ac10b-58cc 3 1 second part
+f47ac10b-58cc 3 2 third part
+```
```
Comment (mavam) [inline src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx:26]: Is this a real-world example? Then we could say what product it is.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+    segment_total: 3,
+    segment_number: 1,
+    payload: "second part",
+  },
+}
+```
+
+:::note[Numeric identifiers]
+If your ids are numeric and zero-padded, or your payloads carry values that must
+survive verbatim—Cisco ISE has both—pass `raw=true` on <Fn>parse_grok</Fn> so
```
Comment (mavam) [inline src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Ah, here we have it: Cisco ISE. Move that up.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+:::caution[Size the span to the segment spread]
+A message's segments only reassemble if they fall close enough in event time to
+share a window. If they can arrive more than `span` apart, raise `span` and use
+the hopping windows shown earlier so they still meet. Sized too small, a message
+comes out split into incomplete pieces.
+:::
+
+## In practice
+
+The Cisco package combines all of these steps for Cisco ISE syslog, which splits
```
Comment (mavam) [inline src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: "The Cisco package": In general, we try to avoid referencing library content because it is still changing quite rapidly. Here, when you mention the Cisco package, it's not quite clear what this actually is. What you could do, however, is make this an aside and mention that we have a package and link to it also on GitHub. Without a link, this section is not helpful. I would move this aside further up, though. The first time you mentioned the Cisco ISE format and mentioned, "Hey, by the way, do you want to check out how this works in practice? Then take a look at our package in the library."

**Review by mavam — APPROVED** [review]
Reads really well!

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx (hunk 5 above)
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Removed every em dash from the guide (was 8, now 0).

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx (hunk 6 above)
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx:12]: Done — 'The problem' now reads: "Cisco Identity Services Engine (ISE), for example, does this with its syslog."

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx (hunk 7 above)
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx:26]: Added after the wire-format block: "This mirrors how Cisco ISE frames its syslog segments."

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx (hunk 8 above)
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Moved up — Cisco ISE is now named in the first paragraph of 'The problem', before the framing example and the raw note.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx (hunk 9 above)
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Done — dropped the bottom "In practice" section and added a `:::tip[See it in practice]` aside right after the wire format that mentions the package and links it on GitHub (tenzir/library/tree/main/cisco).

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+  sort segment_number
+  summarize \
+    message_id,
+    first_segment_window_start=min(segment_window_start),
+    header_window_start=first(header_window_start),
+    payloads=collect(payload)
+  // Emit only from the window anchored at segment 0, or for an orphan run of
+  // continuations, the previous overlapping window.
+  orphan_window_start = first_segment_window_start - $span
+  emit_window_start = header_window_start else orphan_window_start
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Fixed. The operator now uses 3×span hopping windows and emits each multi-segment message from the window centered on segment 0's bucket (reaching a span on either side), so a header just after a boundary still includes a continuation just before it. The guide is re-synced to that, and the earlier "assumes segment 0 is earliest" caveat is removed. Verified with a regression test where the header arrives later than its continuation — it now reassembles complete.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+from {message_id: "f47ac10b", segment_number: 2, ts: 2024-01-01T10:00:31, payload: "c=3"},
+     {message_id: "f47ac10b", segment_number: 0, ts: 2024-01-01T10:00:29, payload: "a=1"},
+     {message_id: "f47ac10b", segment_number: 1, ts: 2024-01-01T10:00:29, payload: "b=2"}
+// Record each segment's fixed window; only segment 0 carries the header.
+segment_window_start = floor(ts, $span)
+header_window_start = segment_window_start if segment_number == 0 else null
+window size=$span * 2, every=$span, on=ts, idle_timeout=$span {
+  sort segment_number
+  summarize \
+    message_id,
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Documented as a constraint: the caution now says to keep `3 * span` (the window width) shorter than the id-reuse interval, since two messages reusing an id inside one window merge. For Cisco ISE specifically, msg_id is a per-node monotonic counter that only repeats on application restart or across multiple nodes (per Cisco's syslog docs), and the single-segment case is already disambiguated by a payload-qualified key; the residual generic multi-segment case is the documented limit.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+```tql
+let $span = 10s
+
+from {message_id: "f47ac10b", segment_number: 2, ts: 2024-01-01T10:00:31, payload: "c=3"},
+     {message_id: "f47ac10b", segment_number: 0, ts: 2024-01-01T10:00:29, payload: "a=1"},
+     {message_id: "f47ac10b", segment_number: 1, ts: 2024-01-01T10:00:29, payload: "b=2"}
+// Record each segment's fixed window; only segment 0 carries the header.
+segment_window_start = floor(ts, $span)
+header_window_start = segment_window_start if segment_number == 0 else null
+window size=$span * 2, every=$span, on=ts, idle_timeout=$span {
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Addressed in the "Size the span" caution: it quotes the exact `window dropped … late event(s)` warning, explains it fires when a straggler's timestamp predates the advanced clock, and says to raise `idle_timeout` and `tolerance` for out-of-order data — a segment delayed past its window cannot be recovered.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+
+```tql
+let $span = 10s
+
+from {message_id: "f47ac10b", segment_number: 2, ts: 2024-01-01T10:00:31, payload: "c=3"},
+     {message_id: "f47ac10b", segment_number: 0, ts: 2024-01-01T10:00:29, payload: "a=1"},
+     {message_id: "f47ac10b", segment_number: 1, ts: 2024-01-01T10:00:29, payload: "b=2"}
+// Record each segment's fixed window; only segment 0 carries the header.
+segment_window_start = floor(ts, $span)
+header_window_start = segment_window_start if segment_number == 0 else null
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx:136]: Good catch. Updated the Numeric identifiers note: with `raw=true`, cast the segment number with `int()` right after framing — before the header check (`segment_number == 0`), the sort, and the completeness count — not just before the sort.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+so type inference doesn't rewrite them (see the `raw` option). Then cast the
+segment number back with `int()` right after framing, since the header check
+(`segment_number == 0`), the sort, and the completeness count all need it numeric.
+:::
+
+## Gather the segments
+
+Bound the segments of one message in event time, order them, and join their
+payloads. Use **hopping windows** (`size=$span * 2, every=$span`) so the windows
+overlap by `span` and any two segments within `span` of each other land in a
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Fixed — the prose now says `size=$span * 3` to match the examples and the guard, and the overlap wording is corrected.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+  // of continuations emits from the window before its first segment.
+  orphan_window_start = first_segment_window_start - $span
+  emit_window_start = orphan_window_start
+  if header_window_start != null {
+    emit_window_start = header_window_start - $span
+  }
+  where $window.start == emit_window_start
+  drop first_segment_window_start, header_window_start, orphan_window_start, emit_window_start
+}
+message = payloads.join(", ")
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Switched the recipe to `join("")` (verbatim concatenation) with the separators living inside the payload chunks, so nothing is inserted between segments. Added a "Join verbatim" note: sources that split on a delimiter and drop it rejoin with that delimiter — Cisco ISE splits on `, `, which is why its operator trims and rejoins with `, `.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+
+Rather than drop a message that lost a segment, count how many distinct segments
+arrived and flag the result, so downstream stages can handle a partial message
+instead of losing it. Collect `segment_total` and `segment_numbers` in the
+window's `summarize`, then compare after it:
+
+```tql
+received = segment_numbers.distinct().where(n => n >= 0 and n < segment_total).length()
+reassembly_complete = received == segment_total
+missing_segments = segment_total - received
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx:226]: Fixed — the note now says to cast the segment number **and total** with `int()` after framing, since the completeness count compares and subtracts `segment_total`.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+:::tip[One window is enough when segments share a timestamp]
+Hopping windows matter only because a segment's `on` value can differ slightly
+between segments of the same message, for example the per-segment syslog receipt
+time. If the message body instead carries its own timestamp that is identical for
+every segment of one sequence (often distinct from the syslog timestamp), window
+on that field: all segments then fall in the same window, and a plain
+`window size=$span` without the boundary machinery reassembles them.
+
+```tql
+window size=$span, on=message_time {
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx]: Added `idle_timeout=$span` to the shared-timestamp snippet, matching the main examples.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+
+## Report completeness
+
+Rather than drop a message that lost a segment, count how many distinct segments
+arrived and flag the result, so downstream stages can handle a partial message
+instead of losing it. Collect `segment_total` and `segment_numbers` in the
+window's `summarize`, then compare after it:
+
+```tql
+received = segment_numbers.distinct().where(n => n >= 0 and n < segment_total).length()
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx:224]: Fixed — the Report completeness section now shows the window's `summarize` producing the inputs first (`segment_total=first(segment_total)` as a scalar and `segment_numbers=collect(segment_number)`), then the after-window comparison.

File: src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+## Drop duplicate segments
+
+A segment can arrive more than once. Add <Op>deduplicate</Op> inside the window,
+right after `sort`, keyed by message and segment number, so a redelivered segment
+contributes its payload only once:
+
+```tql
+window size=$span * 3, every=$span, on=ts, idle_timeout=$span {
+  sort segment_number
+  deduplicate message_id, segment_number
```
[the remark this answers is not in the document]
Comment (zedoraps) [reply src/content/docs/guides/parsing/reassemble-multi-segment-messages.mdx:193]: This is a manifestation of out-of-order event-time arrival rather than something a dedup tie-breaker fixes (a sort-by-bucket tie-breaker doesn't help: the complete window still fails the emit guard). It's now covered in the new "Limitations and trade-offs" section as a known edge case to test with the delay fixture and bound with `tolerance`. The packaged `cisco::ise::reassemble` handles ISE's specific case using segment 0's body timestamp; the generic recipe documents it as a limitation.
